internal/backends/linuxkms/renderer/sw.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright © SixtyFPS GmbH <info@slint.dev>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-Royalty-free-1.1 OR LicenseRef-Slint-commercial
+
+//! Delegate the rendering to the [`i_slint_core::software_renderer::SoftwareRenderer`]
+
+use i_slint_core::api::PhysicalSize as PhysicalWindowSize;
+use i_slint_core::platform::PlatformError;
+pub use i_slint_core::software_renderer::SoftwareRenderer;
+use i_slint_core::software_renderer::{PremultipliedRgbaColor, RepaintBufferType, TargetPixel};
+use std::rc::Rc;
+
+use crate::display::{Presenter, RenderingRotation};
+use crate::drmoutput::DrmOutput;
+
+pub struct SoftwareRendererAdapter {
+    renderer: SoftwareRenderer,
+    display: Rc<crate::display::swdisplay::SoftwareBufferDisplay>,
+    size: PhysicalWindowSize,
+}
+
+#[repr(transparent)]
+#[derive(Copy, Clone, bytemuck::Pod, bytemuck::Zeroable)]
+struct DumbBufferPixel(pub u32);
+
+impl From<DumbBufferPixel> for PremultipliedRgbaColor {
+    #[inline]
+    fn from(pixel: DumbBufferPixel) -> Self {
+        let v = pixel.0;
+        PremultipliedRgbaColor {
+            red: (v >> 16) as u8,
+            green: (v >> 8) as u8,
+            blue: (v >> 0) as u8,
+            alpha: (v >> 24) as u8,
+        }
+    }
+}
+
+impl From<PremultipliedRgbaColor> for DumbBufferPixel {
+    #[inline]
+    fn from(pixel: PremultipliedRgbaColor) -> Self {
+        Self(
+            (pixel.alpha as u32) << 24
+                | ((pixel.red as u32) << 16)
+                | ((pixel.green as u32) << 8)
+                | (pixel.blue as u32),
+        )
+    }
+}
+
+impl TargetPixel for DumbBufferPixel {
+    fn blend(&mut self, color: PremultipliedRgbaColor) {
+        let mut x = PremultipliedRgbaColor::from(*self);
+        x.blend(color);
+        *self = x.into();
+    }
+
+    fn from_rgb(r: u8, g: u8, b: u8) -> Self {
+        Self(0xff000000 | ((r as u32) << 16) | ((g as u32) << 8) | (b as u32))
+    }
+
+    fn background() -> Self {
+        Self(0)
+    }
+}
+
+impl SoftwareRendererAdapter {
+    pub fn new(
+        device_opener: &crate::DeviceOpener,
+    ) -> Result<Box<dyn crate::fullscreenwindowadapter::FullscreenRenderer>, PlatformError> {
+        let drm_output = DrmOutput::new(device_opener)?;
+        let display = Rc::new(crate::display::swdisplay::SoftwareBufferDisplay::new(drm_output)?);
+
+        let (width, height) = display.drm_output.size();
+        let size = i_slint_core::api::PhysicalSize::new(width, height);
+
+        let renderer = Box::new(Self { renderer: SoftwareRenderer::new(), display, size });
+
+        eprintln!("Using Software renderer");
```

Comment:
leftover?

Reply:
No, all the renderers currently print when they’re in use.